uninstall: delete CephCluster at the start of uninstall

jarrpa opened the pull request:
To properly trigger the new uninstallation logic in Rook-Ceph for a
"graceful" uninstall, we need to mark the CephCluster for deletion
towards the start of the process. This is how we inform Rook-Ceph that a
cluster deletion is in progress.

Signed-off-by: Jose A. Rivera <jarrpa@redhat.com>

nb-ohad commented:
@jarrpa Do we know how this will affect external mode / provider-consumer mode?
In provider consumer mode strict ordering is important, let's make sure we are not breaking it

@iamniting Can you please review the code and give your opnion

iamniting commented:
> @jarrpa Do we know how this will affect external mode / provider-consumer mode? In provider consumer mode strict ordering is important, let's make sure we are not breaking it
> 
> @iamniting Can you please review the code and give your opinion

@nb-ohad This change should not break our provider/consumer uninstall.

@jarrpa The ensureDeleted func will always return an error until the cephcluster is present and we won't be able to proceed further at all. If we just want to mark it for deletion we should change the func too.

umangachapagain reviewed (changes requested):
If we do this, CephCluster will be marked for deletion but since finalizer is set, it will not be removed yet. So `ensureDelete` is going to return an error and the deletion loop will never proceed forward until Rook removed finalizer.
For Rook to remove finalizer, all the consumers of CephCluster needs to be deleted but we can't do it since we're waiting on CephCluster to be removed.
Seems like a deadlock. This can be fixed by returning `nil` instead of error when CephCluster is present and deletion timestamp is set. But, then that would mean we'll never be able to guarantee that CephCluster resource is cleaned up.

jarrpa commented:
I verified the above comments, it just took me a while to get a testing cluster up and running.

I reverted the move of the ensureDeleted call and updated the annotating function instead. @travisn's suggestion is ideal, but that's a larger refactor than we really want to do right now.

PTAL

jarrpa commented:
/test ocs-operator-bundle-e2e-aws

jarrpa commented:
/override ci/prow/red-hat-storage-ocs-ci-e2e-aws

iamniting commented on controllers/storagecluster/uninstall_reconciler.go:
```diff
@@ -301,13 +300,20 @@ func (r *StorageClusterReconciler) deleteResources(sc *ocsv1.StorageCluster) (re
 		return reconcile.Result{}, err
 	}
 
-	err = r.setRookUninstallandCleanupPolicy(sc)
```
Why are we removing this func call? This is required to tell rook about the kind of uninstall (force/graceful). We should not remove this.

jarrpa replied:
I pushed the wrong commit, this one was out of date

jarrpa replied:
Also note this moved to the newCephCluster() function. I couldn't figure out why we weren't always setting them to begin with.

jarrpa commented:
/override ci/prow/red-hat-storage-ocs-ci-e2e-aws

jarrpa commented:
/override ci/prow/red-hat-storage-ocs-ci-e2e-aws

iamniting reviewed (approved):
/lgtm

agarwal-mudit commented:
/cherry-pick release-4.10
